spacy_transformers/util.py

```diff
@@ -1,426 +1,214 @@
-    """Pad a batch of lists or 1d arrays with zeros so that sequences are the same
-    length, and form them into a single array.
-    """
-    padded: List[Array] = []
-    seq: Array
-    values = (0, value)
-    pad_desc = [[0, 0]]
-    for seq in batch:
-        pad_desc[0][1] = to - len(seq)
-        padded.append(xp.pad(seq, pad_desc, mode="constant", constant_values=values))
-    output = xp.vstack(padded)
-    assert output.shape == (len(batch), to), output.shape
-    return output
-
-
-def _pad_batch_nd(
-    batch: List[Array], axis: int, *, xp=numpy, to: int = 0, value=-1
-) -> Array:
-    padded: List[Array] = []
-    seq: Array
-    values = (0, value)
-    pad_desc = [[0, 0] for _ in batch[0].shape]
-    for seq in batch:
-        # Ugh, numpy.pad sucks.
-        pad_desc[axis][1] = to - seq.shape[axis]
-        arr = xp.pad(seq, pad_desc, mode="constant", constant_values=values)
-        if len(arr.shape) == 2:
-            # This prevents us concatenating on the sequence dimension, when what
-            # we want is to have a new batch dimension.
-            arr = arr.reshape((1, arr.shape[0], arr.shape[1]))
-        padded.append(arr)
-    output = xp.vstack(padded)
-    return output
-
-
-def batch_by_length(seqs: Union[List[Array]], max_words: int) -> List[List[int]]:
-    """Given a list of sequences, return a batched list of indices into the
-    list, where the batches are grouped by length, in descending order. Batches
-    may be at most max_words in size, defined as max sequence length * size.
-    """
-    # Use negative index so we can get sort by position ascending.
-    lengths_indices = [(len(seq), i) for i, seq in enumerate(seqs)]
-    lengths_indices.sort()
-    batches: List[List[int]] = []
-    batch: List[int] = []
-    for length, i in lengths_indices:
-        # i = -neg_i
-        if not batch:
-            batch.append(i)
-        elif length * (len(batch) + 1) <= max_words:
-            batch.append(i)
-        else:
-            batches.append(batch)
-            batch = [i]
-    if batch:
-        batches.append(batch)
-    # Check lengths match
-    assert sum(len(b) for b in batches) == len(seqs)
-    # Check no duplicates
-    seen = set()
-    for b in batches:
-        seen.update(b)
-    assert len(seen) == len(seqs)
-    batches = [list(sorted(batch)) for batch in batches]
-    batches.reverse()
-    return batches
-
-
-def ensure3d(arr: Array, *, axis: int = 1) -> Array:
-    """Make sure an array is 3d, inserting a dimension at axis if not."""
-    if arr.size == 0:
-        return arr.reshape((0, 0, 0))
-    elif len(arr.shape) == 3:
-        return arr
-    elif len(arr.shape) == 2:
-        return arr.reshape((arr.shape[0], 1, arr.shape[1]))
-    else:
-        raise ValueError(f"Cannot make array 3d. Shape: {arr.shape}")
 
+# TODO: How should we register this?
+spanners = catalogue.create("spacy-transformers", "spanners")
```

Comment:
For it to be used in thinc's config system, I think we need to add it to [thinc's `registry` object](https://github.com/explosion/thinc/blob/master/thinc/config.py#L144) in `config.py`. If we have `registry.create` return a `Decorator` instead of `None`, we can do this ?
```
spanners = registry.create("spanners")
``` 
It would be in the `thinc` namespace then though (unless we generalize that `registry` object further)